### PR TITLE
Add new DDR commands for modularity__Command 3.1 and 3.2

### PR DESCRIPTION
**Pull request by Ailloviee**
Added two commands:

1.  `!dumpmoduleexports [targertModuleAddress]` display all j9packages
exported by the target module 
2.  `!dumpmodulereads [targetModuleAddress]` display all modules that the
target reads
Related issue: [#1859](https://github.com/eclipse/openj9/issues/1859)
Created ModuleHashTable.java to access J9HashTables in modules

Examples: 
!dumpmoduleexports: 
```
com/ibm/oti/reflect                           !j9package 0x00000130550DB1A8
java/util/concurrent/locks                    !j9package 0x00000130550DB1F8
com/sun/net/ssl/internal/www/protocol/https   !j9package 0x00000130550DB248
sun/net/www/protocol/http/ntlm                !j9package 0x00000130550DB298
sun/net/util                                  !j9package 0x00000130550DB2E8
jdk/internal/perf                             !j9package 0x00000130550DB338
java/nio/charset/spi                          !j9package 0x00000130550DB388
java/util/zip                                 !j9package 0x00000130550DB3D8
sun/util/logging                              !j9package 0x00000130550DB428
java/security/spec                            !j9package 0x00000130550DB478
jdk/internal/misc                             !j9package 0x00000130550DB4C8
sun/net/dns                                   !j9package 0x00000130550DB518
java/io                                       !j9package 0x00000130550DB568
java/security                                 !j9package 0x00000130550DB5B8
jdk/internal/org/objectweb/asm                !j9package 0x00000130550DB608
java/util/jar                                 !j9package 0x00000130550DB658
sun/security/x509                             !j9package 0x00000130550DB6A8
java/util/regex                               !j9package 0x00000130550DB6F8
sun/net/spi                                   !j9package 0x00000130550DB748
sun/security/ssl                              !j9package 0x00000130550DA098
com/sun/crypto/provider                       !j9package 0x00000130550DA0E8
sun/reflect/generics/reflectiveObjects        !j9package 0x00000130550DA138
com/ibm/jit                                   !j9package 0x00000130550DA188
jdk/internal/jimage/decompressor              !j9package 0x00000130550DA1D8
com/sun/net/ssl/internal/ssl                  !j9package 0x00000130550DA228
java/nio/channels                             !j9package 0x00000130550DA278
java/util/concurrent                          !j9package 0x00000130550DA2C8
sun/util/resources/cldr                       !j9package 0x00000130550DA318
sun/util/cldr                                 !j9package 0x00000130550DA368
javax/security/auth/x500                      !j9package 0x00000130550DA3B8
sun/security/action                           !j9package 0x00000130550DA408
java/nio                                      !j9package 0x00000130550DA458
sun/text/spi                                  !j9package 0x00000130550DA4A8
java/lang                                     !j9package 0x00000130550DA4F8
jdk/internal/org/objectweb/asm/tree/analysis  !j9package 0x00000130550DA548
javax/security/cert                           !j9package 0x00000130550DA598
sun/net/ext                                   !j9package 0x00000130550DA5E8
sun/net/www/protocol/jrt                      !j9package 0x00000130550DA638
java/security/acl                             !j9package 0x00000130550DA688
sun/text/resources/cldr                       !j9package 0x00000130550DA6D8
sun/net/www/protocol/mailto                   !j9package 0x00000130550DA728
sun/security/rsa                              !j9package 0x00000130550DA778
java/util                                     !j9package 0x00000130550DA7C8
java/text/spi                                 !j9package 0x00000130550DA818
sun/util/locale                               !j9package 0x00000130550DA868
sun/security/tools                            !j9package 0x00000130550DA8B8
java/nio/file/attribute                       !j9package 0x00000130550DA908
jdk/internal/org/xml/sax                      !j9package 0x00000130550DA958
javax/security/auth/spi                       !j9package 0x00000130550DA9A8
jdk/internal/ref                              !j9package 0x00000130550DA9F8
sun/invoke/util                               !j9package 0x00000130550DAA48
sun/security/pkcs                             !j9package 0x00000130550DAA98
java/nio/file/spi                             !j9package 0x00000130550DAAE8
com/ibm/oti/vm                                !j9package 0x00000130550DAB38
java/nio/channels/spi                         !j9package 0x00000130550DAB88
jdk/internal/org/xml/sax/helpers              !j9package 0x00000130550DABD8
jdk/internal/loader                           !j9package 0x00000130550DAC28
sun/nio/fs                                    !j9package 0x00000130550DAC78
jdk/internal/org/objectweb/asm/tree           !j9package 0x00000130550DACC8
sun/util/calendar                             !j9package 0x00000130550DAD18
javax/security/auth/login                     !j9package 0x00000130550DAD68
java/nio/charset                              !j9package 0x00000130550DADB8
com/ibm/oti/util                              !j9package 0x00000130550DAE08
javax/crypto/interfaces                       !j9package 0x00000130550DAE58
sun/reflect/generics/factory                  !j9package 0x00000130550DAEA8
sun/util/spi                                  !j9package 0x00000130550DAEF8
java/net/spi                                  !j9package 0x00000130550DAF48
sun/security/timestamp                        !j9package 0x00000130550DAF98
sun/text/normalizer                           !j9package 0x00000130550DAFE8
sun/net/www/protocol/jmod                     !j9package 0x00000130550DF5E8
sun/net/www/protocol/http                     !j9package 0x00000130550DF638
com/sun/net/ssl                               !j9package 0x00000130550DF688
sun/reflect/generics/tree                     !j9package 0x00000130550DF6D8
jdk/internal/vm/annotation                    !j9package 0x00000130550DF728
java/security/cert                            !j9package 0x00000130550DF778
javax/crypto                                  !j9package 0x00000130550DF7C8
sun/reflect/generics/repository               !j9package 0x00000130550DF818
sun/net/smtp                                  !j9package 0x00000130550DF868
com/sun/java/util/jar/pack                    !j9package 0x00000130550DF8B8
sun/reflect/generics/scope                    !j9package 0x00000130550DF908
com/ibm/sharedclasses/spi                     !j9package 0x00000130550DF958
jdk/internal/org/objectweb/asm/commons        !j9package 0x00000130550DF9A8
sun/net                                       !j9package 0x00000130550DF9F8
sun/launcher/resources                        !j9package 0x00000130550DFA48
com/ibm/tools/attach/target                   !j9package 0x00000130550DFA98
com/ibm/gpu/spi                               !j9package 0x00000130550DFAE8
sun/util/locale/provider                      !j9package 0x00000130550DFB38
sun/util/resources                            !j9package 0x00000130550DFB88
java/time/zone                                !j9package 0x00000130550DFBD8
sun/reflect/generics/visitor                  !j9package 0x00000130550DFC28
jdk/internal/jmod                             !j9package 0x00000130550DFC78
sun/security/pkcs12                           !j9package 0x00000130550DFCC8
com/sun/security/cert/internal/x509           !j9package 0x00000130550DFD18
sun/security/pkcs10                           !j9package 0x00000130550DFD68
sun/net/ftp/impl                              !j9package 0x00000130550DFDB8
sun/security/provider/certpath                !j9package 0x00000130550DFE08
sun/security/provider/certpath/ssl            !j9package 0x00000130550DFE58
sun/security/util                             !j9package 0x00000130550DFEA8
java/security/interfaces                      !j9package 0x00000130550DFEF8
java/lang/ref                                 !j9package 0x00000130550DFF48
sun/reflect/misc                              !j9package 0x00000130550DFF98
java/text                                     !j9package 0x00000130550DFFE8
sun/nio                                       !j9package 0x00000130550E0038
sun/io                                        !j9package 0x00000130550E0088
sun/net/www/protocol/file                     !j9package 0x00000130550E00D8
jdk/internal/jrtfs                            !j9package 0x00000130550E0128
sun/security/internal/spec                    !j9package 0x00000130550E0178
sun/net/www                                   !j9package 0x00000130550E01C8
sun/reflect/annotation                        !j9package 0x00000130550E0218
java/time/chrono                              !j9package 0x00000130550E0268
jdk/internal/org/objectweb/asm/util           !j9package 0x00000130550E02B8
sun/security/tools/keytool                    !j9package 0x00000130550E0308
sun/net/www/protocol/https                    !j9package 0x00000130550E0358
sun/security/validator                        !j9package 0x00000130550E03A8
java/net                                      !j9package 0x00000130550E03F8
sun/net/www/protocol/jar                      !j9package 0x00000130550E0448
jdk/internal/util                             !j9package 0x00000130550E0498
jdk/internal/util/jar                         !j9package 0x00000130550E04E8
com/ibm/oti/lang                              !j9package 0x00000130550E0538
java/time                                     !j9package 0x00000130550C7F28
sun/security/internal/interfaces              !j9package 0x00000130550C7F78
sun/text                                      !j9package 0x00000130550C7FC8
java/time/temporal                            !j9package 0x00000130550C8018
javax/security/auth/callback                  !j9package 0x00000130550C8068
java/lang/reflect                             !j9package 0x00000130550C80B8
sun/text/bidi                                 !j9package 0x00000130550C8108
java/util/function                            !j9package 0x00000130550C8158
sun/text/resources                            !j9package 0x00000130550C81A8
java/lang/invoke                              !j9package 0x00000130550C81F8
javax/net/ssl                                 !j9package 0x00000130550C8248
java/lang/module                              !j9package 0x00000130550C8298
jdk/internal/math                             !j9package 0x00000130550C82E8
java/nio/file                                 !j9package 0x00000130550C8338
jdk/internal/reflect                          !j9package 0x00000130550C8388
sun/reflect/generics/parser                   !j9package 0x00000130550C83D8
sun/util                                      !j9package 0x00000130550C8428
jdk/internal/util/xml/impl                    !j9package 0x00000130550C8478
sun/net/www/content/text                      !j9package 0x00000130550C84C8
sun/net/sdp                                   !j9package 0x00000130550C8518
sun/net/www/protocol/ftp                      !j9package 0x00000130550C8568
java/util/spi                                 !j9package 0x00000130550C85B8
sun/security/provider                         !j9package 0x00000130550C8608
javax/security/auth                           !j9package 0x00000130550C8658
com/sun/security/ntlm                         !j9package 0x00000130550C86A8
sun/net/idn                                   !j9package 0x00000130550C86F8
sun/invoke/empty                              !j9package 0x00000130550C8748
jdk/internal                                  !j9package 0x00000130550C8798
jdk/internal/org/objectweb/asm/signature      !j9package 0x00000130550C87E8
jdk/internal/util/xml                         !j9package 0x00000130550C8838
jdk/internal/logger                           !j9package 0x00000130550C8888
sun/net/ftp                                   !j9package 0x00000130550C88D8
javax/crypto/spec                             !j9package 0x00000130550C8928
jdk/internal/jimage                           !j9package 0x00000130550C8978
java/time/format                              !j9package 0x00000130550C89C8
sun/launcher                                  !j9package 0x00000130550C8A18
javax/net                                     !j9package 0x00000130550C8A68
java/lang/annotation                          !j9package 0x00000130550C8AB8
sun/invoke                                    !j9package 0x00000130550C8B08
jdk/internal/module                           !j9package 0x00000130550C8B58
sun/nio/cs                                    !j9package 0x00000130550C8BA8
sun/security/jca                              !j9package 0x00000130550C8BF8
java/util/concurrent/atomic                   !j9package 0x00000130550C8C48
sun/nio/ch                                    !j9package 0x00000130550C8C98
java/util/stream                              !j9package 0x00000130550C8CE8
java/math                                     !j9package 0x00000130550C8D38
jdk/internal/vm                               !j9package 0x00000130550C8D88
sun/net/www/http                              !j9package 0x00000130550C8DD8
```
!dumpmodulereads: 
```
> !dumpmodulereads 0x00000130550C7E88
java.instrument                !j9module 0x000001305F46DB38
jdk.xml.dom                    !j9module 0x00000130550DBB58
jdk.accessibility              !j9module 0x000001305F469EA8
java.logging                   !j9module 0x000001305F477388
jdk.jshell                     !j9module 0x000001305F471D98
jdk.jdwp.agent                 !j9module 0x000001305F477248
openj9.dtfj                    !j9module 0x000001305F469FE8
java.scripting                 !j9module 0x000001305F46DD98
jdk.dynalink                   !j9module 0x000001305F4685E8
jdk.jdeps                      !j9module 0x000001305F469518
openj9.zosconditionhandling    !j9module 0x000001305F4710C8
jdk.management.agent           !j9module 0x000001305F469298
jdk.internal.opt               !j9module 0x000001305F46ABA8
openj9.jvm                     !j9module 0x000001305F471528
java.compiler                  !j9module 0x00000130550DBCE8
java.sql                       !j9module 0x00000130550DB798
openj9.gpu                     !j9module 0x00000130550DB978
java.xml.crypto                !j9module 0x000001305F46CDC8
jdk.naming.rmi                 !j9module 0x000001305F477D68
java.naming                    !j9module 0x000001305F477EF8
openj9.cuda                    !j9module 0x000001305F477298
jdk.jdi                        !j9module 0x00000130550DBF18
java.xml                       !j9module 0x000001305F46BCB8
openj9.sharedclasses           !j9module 0x00000130550DBA68
jdk.attach                     !j9module 0x000001305F478798
java.desktop                   !j9module 0x000001305F46E1F8
jdk.jartool                    !j9module 0x000001305F46B6E8
jdk.crypto.ec                  !j9module 0x000001305F477E58
jdk.localedata                 !j9module 0x000001305F471168
jdk.unsupported                !j9module 0x000001305F4684A8
jdk.naming.dns                 !j9module 0x000001305F46CCD8
jdk.jsobject                   !j9module 0x000001305F46DA48
jdk.compiler                   !j9module 0x000001305F46ACE8
java.security.sasl             !j9module 0x000001305F46DED8
jdk.scripting.nashorn          !j9module 0x000001305F4767A8
openj9.dataaccess              !j9module 0x000001305F477CC8
java.prefs                     !j9module 0x000001305F478928
java.management                !j9module 0x000001305F468B18
jdk.charsets                   !j9module 0x000001305F471398
jdk.net                        !j9module 0x000001305F471CF8
jdk.security.auth              !j9module 0x000001305F478BF8
jdk.editpad                    !j9module 0x000001305F478D88
jdk.crypto.cryptoki            !j9module 0x000001305F478E78
jdk.jlink                      !j9module 0x000001305F4699A8
java.rmi                       !j9module 0x000001305F478F68
java.smartcardio               !j9module 0x000001305F471438
jdk.internal.le                !j9module 0x000001305F46B8C8
jdk.internal.ed                !j9module 0x000001305F4691A8
jdk.crypto.mscapi              !j9module 0x000001305F469478
java.security.jgss             !j9module 0x000001305F471708
jdk.management                 !j9module 0x000001305F478448
jdk.httpserver                 !j9module 0x000001305F479468
jdk.internal.jvmstat           !j9module 0x000001305F4795A8
jdk.zipfs                      !j9module 0x00000130550DB8D8
jdk.javadoc                    !j9module 0x000001305F474498
jdk.jstatd                     !j9module 0x000001305F4715C8
jdk.sctp                       !j9module 0x000001305F46D868
jdk.security.jgss              !j9module 0x000001305F46D958
java.management.rmi            !j9module 0x000001305F46E0B8
jdk.jconsole                   !j9module 0x000001305F469D18
java.sql.rowset                !j9module 0x000001305F4789C8
java.se                        !j9module 0x000001305F46DE88
java.datatransfer              !j9module 0x000001305F476668
```

Signed-off-by: Charles_Zheng <Juntian.Zheng@ibm.com>

### CODE REVIEW

**Comment by Ailloviee** [comment]
-fyi @tajila

**Comment by tajila** [comment]
!dumppackageexports should print the packageName then the `!j9package 0x00000000XXXXXX`

Same with !dumpmodulereads

**Comment by AdamBrousseau** [comment]
Jenkins signed off by check

**Comment by tajila** [comment]
@keithc-ca can you please review these changes

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ModuleHashTable.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	 * @param structure The J9HashTablePointer
+	 * @throws CorruptDataException When fails to read from structure
+	 * @return The module hash table opened from the given j9 hash table pointer
+	 */
+	public static HashTable<J9ModulePointer> fromJ9HashTable(J9HashTablePointer structure) throws CorruptDataException 
+	{
+		return new ModuleHashTable(structure, false, J9ModulePointer.class, new ModuleHashEqualFn(), new ModuleHashFn());
+	}
+	
+	private static class ModuleHashFn implements HashFunction<J9ModulePointer>
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ModuleHashTable.java]: These private nested classes should each have an explicit (package-visible) constructor to avoid the need for synthetic accessor methods.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpModuleReadsCommand.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+					String hexAddress = readModulePtr.getHexAddress();
+					out.printf("%-30s !j9module %s%n", moduleName, hexAddress);
+				}
+			}
+		} catch (CorruptDataException e) {
+			throw new DDRInteractiveCommandException(e);
+		}
+	}
+}
+
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpModuleReadsCommand.java]: Please remove the extra blank line here and in other changed files.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckEngine.java [hunk 3]
```diff
@@ -1463,3 +1463,4 @@ public boolean needVerifyOwnableSynchronizerConsistency()
 		return _needVerifyOwnableSynchronizerConsistency;
 	}
 }
+
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckEngine.java]: After removing the blank line, this file would be unchanged. Please remove it from the PR.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpPackageExportsCommand.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+/**
+ * DumpPackageExports command displays all display all j9packages exported by the target module
+ * 
+ * Example:
+ * 			!dumppackageexports 0x00000130550C7E88
+ * Example output: 
+ * 			!j9package 0x00000130550DB1A8
+ *			!j9package 0x00000130550DB1F8
+ *			!j9package 0x00000130550DB248
+ *			.....(A list of every package the module exports)
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpPackageExportsCommand.java]: Please avoid using tabs in places other at the beginning of a line.
The examples need not be indented so much.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpModuleReadsCommand.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+ * 			!j9module 0x000001305F46DB38
+ *			!j9module 0x00000130550DBB58
+ *			!j9module 0x000001305F469EA8
+ *			.....(A list of every module the targeted module reads)
+ */
+public class DumpModuleReadsCommand extends Command{
+
+	public DumpModuleReadsCommand()
+	{
+		addCommand("dumpmodulereads", "<targetModuleAddress>", "display all modules that the target module reads");
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpModuleReadsCommand.java]: This would read better as: 'Display all modules read by the target module'.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ModuleHashTable.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	public static HashTable<J9ModulePointer> fromJ9HashTable(J9HashTablePointer structure) throws CorruptDataException 
+	{
+		return new ModuleHashTable(structure, false, J9ModulePointer.class, new ModuleHashEqualFn(), new ModuleHashFn());
+	}
+	
+	private static class ModuleHashFn implements HashFunction<J9ModulePointer> 
+	{
+		public UDATA hash(J9ModulePointer entry) throws CorruptDataException 
+		{
+			StringHashFunction<J9ModulePointer> stringHashFn = new StringTable.StringHashFunction<J9ModulePointer>();
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ModuleHashTable.java]: Can the hash (and equal) functions be reused?

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/JavaVersionHelper.java [hunk 7]
```diff
@@ -54,3 +54,4 @@ public static boolean ensureJava9AndUp(J9JavaVMPointer vm, PrintStream out) thro
 		return true;
 	}
 }
+
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/JavaVersionHelper.java]: Please undo this change.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpModuleReadsCommand.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+		}
+		try {
+			J9JavaVMPointer vm = J9RASHelper.getVM(DataType.getJ9RASPointer());
+			if (JavaVersionHelper.ensureJava9AndUp(vm, out)) {
+				String targetModuleAddress = args[0];
+				J9ModulePointer modulePtr = J9ModulePointer.cast(Long.decode(targetModuleAddress));
+				J9HashTablePointer readTable = modulePtr.readAccessHashTable();
+				HashTable<J9ModulePointer> moduleHashTable = ModuleHashTable.fromJ9HashTable(readTable);
+				SlotIterator<J9ModulePointer> slotIterator = moduleHashTable.iterator();
+				J9ModulePointer readModulePtr = null;
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpModuleReadsCommand.java]: Please move this declaration within the loop below.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ModuleHashTable.java (hunk 6 above)
Comment (Ailloviee) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ModuleHashTable.java]: The two functions are only for J9Module hash tables I think, called by the constructor.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ModuleHashTable.java (hunk 1 above)
Comment (Ailloviee) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ModuleHashTable.java]: I'm not sure what I should put in the constructor..

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ModuleHashTable.java (hunk 6 above)
Comment (keithc-ca) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ModuleHashTable.java]: Yes, but every time `hash` is called, it creates a new `StringHashFunction` object: is that necessary or can that object be reused?

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ModuleHashTable.java (hunk 1 above)
Comment (keithc-ca) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ModuleHashTable.java]: Just a call to `super();`. The point is that the compiler will automatically generate a no-argument constructor with the same visibility as the class (private in this case). That constructor can't be called directly and a synthetic accessor is thus created. If you make the constructor explicit it can have non-private visibility.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpPackageExportsCommand.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+import com.ibm.j9ddr.vm29.pointer.generated.J9JavaVMPointer;
+import com.ibm.j9ddr.vm29.pointer.generated.J9ModulePointer;
+import com.ibm.j9ddr.vm29.pointer.generated.J9PackagePointer;
+import com.ibm.j9ddr.vm29.pointer.generated.J9PoolPointer;
+import com.ibm.j9ddr.vm29.pointer.helper.J9RASHelper;
+import com.ibm.j9ddr.vm29.pointer.helper.J9UTF8Helper;
+import com.ibm.j9ddr.vm29.tools.ddrinteractive.JavaVersionHelper;
+
+/**
+ * DumpPackageExports command displays all display all j9packages exported by the target module
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpPackageExportsCommand.java]: Might a better name for this command be 'dumpmoduleexports'?

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/structureformat/extensions/J9ModuleStructureFormatter.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import com.ibm.j9ddr.vm29.pointer.helper.J9ObjectHelper;
+
+/**
+ * Structure Formatter that adds a suffix like this:
+ * 
+ * Module name: java.instrument
+ * To display all j9packages exported by the module, use !dumppackageexports 0x000001305F46DB38
+ * To display all modules that the target module reads, use !dumpmodulereads 0x000001305F46DB38
+ * 
+ * When a !j9module is formatted
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/structureformat/extensions/J9ModuleStructureFormatter.java]: Should there more to this sentence/paragraph?

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpPackageExportsCommand.java (hunk 9 above)
Comment (Ailloviee) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpPackageExportsCommand.java]: This command dumps out packages based on module address, hence the name. For `dumpmoduleexports` there is another command: https://github.com/eclipse/openj9/pull/2058 , please see command 5 (the second one in the pr). The command dumps out modules instead based on package address

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/structureformat/extensions/J9ModuleStructureFormatter.java (hunk 10 above)
Comment (Ailloviee) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/structureformat/extensions/J9ModuleStructureFormatter.java]: There isn't.. I used [J9ClassStructureFormatter.java](https://github.com/Ailloviee/openj9/blob/DDRCommands_v3/debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/structureformat/extensions/J9ClassStructureFormatter.java) as template for this one. If you wish, I can make the comments more specified to both files. (The suffix shows up when we run `!j9module $moduleAddress$`)

Comment (keithc-ca) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/structureformat/extensions/J9ModuleStructureFormatter.java]: Please do fix the comment in this file. Poor comments in other files can be dealt with separately.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpPackageExportsCommand.java (hunk 9 above)
Comment (keithc-ca) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/DumpPackageExportsCommand.java]: I think this command should be `dumpmoduleexports` and the other one in #2058 should be `dumpmoduledirectedexports` (or something like that - although the name is rather long) to make the distinction between packages which are universally exported and those which are exported to a limited set of modules).

**Comment by keithc-ca** [comment]
Jenkins test extended plinux

**Comment by keithc-ca** [comment]
Jenkins test extended plinux

**Comment by keithc-ca** [comment]
Jenkins copyright check

**Comment by pshipton** [comment]
Jenkins copyright check

**Comment by pshipton** [comment]
Jenkins line endings check

**Comment by pshipton** [comment]
Jenkins line endings check

**Comment by keithc-ca** [comment]
Jenkins test extended plinux
